### ISSUE TITLE
Mono broke xunit AppDomain marshalling

### ISSUE THREAD

**Issue by jaredpar** [mono/mono#16373]
This bug is fairly serious. It's blocking our standard CI runs. Temporarily disabling Mono until we can get this resolved. 

## Steps to Reproduce

1. Enlist in https://github.com/dotnet/roslyn
2. Run restore.sh
3. Run build.sh
4. Run `test.sh --testMono`

### Current Behavior

The Roslyn tests which engage `AppDomain` remoting are all failing with the following:

```
    [FATAL ERROR] System.InvalidOperationException
      System.InvalidOperationException : Type 'System.Object' of type 'System.RuntimeType' does not derive from System.Object
      Stack Trace:
        
        
        Server stack trace: 
          at System.Runtime.Serialization.FormatterServices.GetParentTypes (System.RuntimeType parentType, System.RuntimeType[]& parentTypes, System.Int32& parentTypeCount) [0x0003f] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.FormatterServices.InternalGetSerializableMembers (System.RuntimeType type) [0x00089] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.FormatterServices+<>c__DisplayClass9_0.<GetSerializableMembers>b__0 (System.Runtime.Serialization.MemberHolder _) [0x00000] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].GetOrAdd (TKey key, System.Func`2[T,TResult] valueFactory) [0x00034] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.FormatterServices.GetSerializableMembers (System.Type type, System.Runtime.Serialization.StreamingContext context) [0x0005e] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitMemberInfo () [0x0003d] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize (System.Object obj, System.Runtime.Serialization.ISurrogateSelector surrogateSelector, System.Runtime.Serialization.StreamingContext context, System.Runtime.Serialization.Formatters.Binary.SerObjectInfoInit serObjectInfoInit, System.Runtime.Serialization.IFormatterConverter converter, System.Runtime.Serialization.Formatters.Binary.ObjectWriter objectWriter, System.Runtime.Serialization.SerializationBinder binder) [0x00158] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize (System.Object obj, System.Runtime.Serialization.ISurrogateSelector surrogateSelector, System.Runtime.Serialization.StreamingContext context, System.Runtime.Serialization.Formatters.Binary.SerObjectInfoInit serObjectInfoInit, System.Runtime.Serialization.IFormatterConverter converter, System.Runtime.Serialization.Formatters.Binary.ObjectWriter objectWriter, System.Runtime.Serialization.SerializationBinder binder) [0x00006] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Write (System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo objectInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo memberNameInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo typeNameInfo) [0x00101] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.WriteArrayMember (System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo objectInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo arrayElemTypeNameInfo, System.Object data) [0x0015a] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.WriteArray (System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo objectInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo memberNameInfo, System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo memberObjectInfo) [0x0025f] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Write (System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo objectInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo memberNameInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo typeNameInfo) [0x00056] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize (System.Object graph, System.Runtime.Remoting.Messaging.Header[] inHeaders, System.Runtime.Serialization.Formatters.Binary.__BinaryWriter serWriter, System.Boolean fCheck) [0x001d3] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph, System.Runtime.Remoting.Messaging.Header[] headers, System.Boolean fCheck) [0x0006e] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph, System.Runtime.Remoting.Messaging.Header[] headers) [0x00000] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph) [0x00000] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at System.Runtime.Remoting.RemotingServices.SerializeCallData (System.Object obj) [0x00053] in <59b1fe55a7ce47ca99c88b42a0622585>:0 
          at (wrapper xdomain-dispatch) Xunit.Sdk.TestMethodTestCase.get_Traits(object,byte[]&,byte[]&)
        
        Exception rethrown at [0]: 
          at (wrapper xdomain-invoke) Xunit.Sdk.TestMethodTestCase.get_Traits()
          at MessageSinkMessageExtensions.Dispatch[TMessage] (Xunit.Abstractions.IMessageSinkMessage message, System.Collections.Generic.HashSet`1[T] messageTypes, Xunit.MessageHandler`1[TMessage] callback) [0x0001a] in C:\projects\xunit\src\xunit.runner.utility\Extensions\MessageSinkMessageExtensions.cs:39 
          at (wrapper remoting-invoke-with-check) Xunit.MessageSinkAdapter.OnMessageWithTypes(Xunit.Abstractions.IMessageSinkMessage,System.Collections.Generic.HashSet`1<string>)
          at (wrapper xdomain-dispatch) Xunit.MessageSinkAdapter.OnMessageWithTypes(object,byte[]&,byte[]&)
```

### Expected Behavior

These remoting operations should succeed. 

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

Mono nightly 6.7.0.235-0nightly1

**Comment by CoffeeFlux** [comment]
Seems the issue appeared somewhere between 02126de318a and 9a5ae23f7e3, with the latter being the final commit included in 6.7.0.235-0nightly1.

**Comment by lambdageek** [comment]
Having some trouble getting this to reproduce on OSX.

Roslyn dotnet/roslyn@919373550e using [Mono 6.7.0.235 on osx](https://xamjenkinsartifact.azureedge.net/build-package-osx-mono/master/4862/9a5ae23f7e3d49b0b32ae423513d2344a837d698/MonoFramework-MDK-master@9a5ae23f7e3-dirty-6.7.0.235.macos10.xamarin.universal.pkg) I don't see any appdomain crashes.

**Comment by lambdageek** [comment]
Exact command line from my machine that to run a testsuite that shows the issue:
```
/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono --debug "/Users/alklig/.nuget/packages/xunit.runner.console/2.4.1-pre.build.4059/tools/net472/xunit.console.exe" "/Users/alklig/work/roslyn/roslyn/artifacts/bin/Microsoft.CodeAnalysis.CSharp.Emit.UnitTests/Debug/net472/Microsoft.CodeAnalysis.CSharp.Emit.UnitTests.dll" -noshadow -xml "/Users/alklig/work/roslyn/roslyn/artifacts/TestResults/Debug/Microsoft.CodeAnalysis.CSharp.Emit.UnitTests_net472_x64.xml"
```

**Comment by lambdageek** [comment]
This looks familiar.  I think we had this problem before.

My best guess for what's happening:
When an xUnit test is failing, xUnit grabs a stack trace in the domain where the test is running and wants to ship it over to the domain of the test runner.  Something in the stack trace data (a RuntimeType?) doesn't get deserialized properly.

**Comment by CoffeeFlux** [comment]
It's looking like https://github.com/mono/mono/commit/f2d813c90df3a6c4d15eab240b0a68ec055c8f5b is the guilty commit here.

**Comment by lambdageek** [comment]
We've been able to reproduce this on the 2019-08 branch, too, and bisected it down to 64f9c2de6f3.  Current theory is that it has to do with corrupted zlib state while uncompressing embedded ppdbs.

I've been running tests with dotnet/roslyn@5551fad965 and [6.6.0.22 (2019-08) 64f9c2de6f3](https://jenkins.mono-project.com/view/Per-commit%20packages/job/build-package-osx-mono/job/2019-08/24/)  using this command line:

```
/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono --debug "/Users
/alklig/.nuget/packages/xunit.runner.console/2.4.1-pre.build.4059/tools/net472/xunit.cons
ole.exe" "/Users/alklig/work/roslyn/roslyn/artifacts/bin/Microsoft.CodeAnalysis.CSharp.Em
it.UnitTests/Debug/net472/Microsoft.CodeAnalysis.CSharp.Emit.UnitTests.dll" -noshadow -pa
rallel none -verbose -xml "/Users/alklig/work/roslyn/roslyn/artifacts/TestResults/Debug/M
icrosoft.CodeAnalysis.CSharp.Emit.UnitTests_net472_x64.xml" > /tmp/aleksey-full.log 2>&1
```

We always see a failure like this before an appdomain failure:
```
    Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.UnsafeTests.PointerArrayForeachEnumerable [FAIL]
      System.IndexOutOfRangeException : Index was outside the bounds of the array.
      Stack Trace:
          at System.Diagnostics.StackTrace.AddFrames (System.Text.StringBuilder sb, System.Boolean separator, System.Boolean& isAsync) [0x00107] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:246 
          at System.Diagnostics.StackTrace.ToString () [0x0005d] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:381 
          at System.Diagnostics.StackTrace.ToString (System.Diagnostics.StackTrace+TraceFormat traceFormat) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:389 
          at System.Environment.GetStackTrace (System.Exception e, System.Boolean needFileInfo) [0x00014] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Environment.cs:1048 
          at System.Exception.GetObjectData (System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context) [0x00020] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/exception.cs:611 
          at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize (System.Object obj, System.Runtime.Serialization.ISurrogateSelector surrogateSelector, System.Runtime.Serialization.StreamingContext context, System.Runtime.Serialization.Formatters.Binary.SerObjectInfoInit serObjectInfoInit, System.Runtime.Serialization.IFormatterConverter converter, System.Runtime.Serialization.Formatters.Binary.ObjectWriter objectWriter, System.Runtime.Serialization.SerializationBinder binder) [0x00128] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryobjectinfo.cs:159 
          at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize (System.Object obj, System.Runtime.Serialization.ISurrogateSelector surrogateSelector, System.Runtime.Serialization.StreamingContext context, System.Runtime.Serialization.Formatters.Binary.SerObjectInfoInit serObjectInfoInit, System.Runtime.Serialization.IFormatterConverter converter, System.Runtime.Serialization.Formatters.Binary.ObjectWriter objectWriter, System.Runtime.Serialization.SerializationBinder binder) [0x00006] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryobjectinfo.cs:110 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Write (System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo objectInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo memberNameInfo, System.Runtime.Serialization.Formatters.Binary.NameInfo typeNameInfo) [0x00101] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryobjectwriter.cs:422 
          at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize (System.Object graph, System.Runtime.Remoting.Messaging.Header[] inHeaders, System.Runtime.Serialization.Formatters.Binary.__BinaryWriter serWriter, System.Boolean fCheck) [0x001d3] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryobjectwriter.cs:190 
          at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph, System.Runtime.Remoting.Messaging.Header[] headers, System.Boolean fCheck) [0x0006e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:233 
          at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph, System.Runtime.Remoting.Messaging.Header[] headers) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:211 
          at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/serialization/formatters/binary/binaryformatter.cs:203 
          at System.Runtime.Remoting.RemotingServices.SerializeExceptionData (System.Exception ex) [0x0001a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Runtime.Remoting/RemotingServices.cs:822 
          at (wrapper xdomain-dispatch) Roslyn.Test.Utilities.Desktop.RuntimeAssemblyManager.Execute(object,byte[]&,byte[]&,string,string[],string&)
          at (wrapper xdomain-invoke) Roslyn.Test.Utilities.Desktop.RuntimeAssemblyManager.Execute(string,string[],System.Nullable`1<int>,string&)
          at (wrapper remoting-invoke-with-check) Roslyn.Test.Utilities.Desktop.RuntimeAssemblyManager.Execute(string,string[],System.Nullable`1<int>,string&)
          at Roslyn.Test.Utilities.Desktop.DesktopRuntimeEnvironment.Execute (System.String moduleName, System.String[] args, System.String expectedOutput) <0x12d793170 + 0x001ba> in <b284a514e6ca4744974d5a1e57c3ba63>:0 
          at Microsoft.CodeAnalysis.Test.Utilities.CompilationVerifier.Emit (System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, System.Collections.Generic.IEnumerable`1[T] manifestResources, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification peVerify, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures) <0x12abdb5d0 + 0x00176> in <b284a514e6ca4744974d5a1e57c3ba63>:0 
          at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.Emit (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Collections.Generic.IEnumerable`1[T] manifestResources, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) <0x12abdb300 + 0x0015a> in <b284a514e6ca4744974d5a1e57c3ba63>:0 
          at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.CompileAndVerifyCommon (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) <0x12abdad90 + 0x00262> in <b284a514e6ca4744974d5a1e57c3ba63>:0 
          at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerify (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] validator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) <0x12abdaaf0 + 0x000d2> in <1ff7c5f0daa540739c5948e60fcd3f36>:0 
          at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerifyException[T] (Microsoft.CodeAnalysis.CSharp.CSharpCompilation comp, System.String expectedMessage, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) <0x12e0ee070 + 0x0010a> in <1ff7c5f0daa540739c5948e60fcd3f36>:0 
          at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerifyException[T] (System.String source, System.String expectedMessage, System.Boolean allowUnsafe, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) <0x12e0ec630 + 0x00112> in <1ff7c5f0daa540739c5948e60fcd3f36>:0 
          at Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.UnsafeTests.PointerArrayForeachEnumerable () <0x12e0ec5b0 + 0x0006a> in <9bf8602f270441c2a6d1458cb62aa3f6>:0 
          at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
          at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-08/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395 
```

**Comment by jaykrell** [comment]
Is the fix to use the z_ prefix and always statically link zlib? Kinda like it used to??

**Comment by lambdageek** [comment]
I think that might be the right bandaid (we'll see).  But it's not a _fix_ in the sense that we don't really understand _why_ it has anything to do with the exceptions that we see.

**Comment by jaykrell** [comment]
Agreed. zlib is uses explicit state passing so is naturally thread safe, right?

**Comment by jaykrell** [comment]
Agreed. zlib is uses explicit state passing so is naturally thread safe, right?

**Comment by CoffeeFlux** [comment]
Forcing the bundled zlib does indeed solve the issue - the suite is reliably passing for me now. Tracking down what exactly is causing the difference may prove difficult though. I'll probably try and isolate some series of tests that causes the issue and then try to hunt down exactly what the difference is between system and bundled zlib.

I think we can push a commit to master that'll force bundled zlib and have Roslyn's test infrastructure working again while we identify a proper fix. Unfortunately, this won't work for Android, so backporting is not an option.

**Comment by jaykrell** [comment]
Why doesn't that work for Android?

**Comment by marek-safar** [comment]
@CoffeeFlux did you manage to track down the underlying issue?

**Comment by CoffeeFlux** [comment]
Still working on it - priority 1 for this week.

**Comment by marek-safar** [comment]
@jaredpar the issue should be resolved in recent master versions.

**Comment by akoeplinger** [comment]
@marek-safar shouldn't we keep this open to track the real fix?

**Comment by marek-safar** [comment]
@akoeplinger we should open a new issue (or reopen the original one if there was any)

**Comment by akoeplinger** [comment]
@CoffeeFlux can you please open a new issue with the summary of the findings?

**Comment by CoffeeFlux** [comment]
Will open a new issue.

Additionally, new Ubuntu nightlies are currently being published (cc @directhex) with the fix that unblocks Roslyn.
